File: domain/src/main/java/com/github/il4enkodev/househeating/domain/interactor/metering/CurrentMetering.java
package com.github.il4enkodev.househeating.domain.interactor.metering;

import com.github.il4enkodev.househeating.domain.entity.metering.Metering;
import com.github.il4enkodev.househeating.domain.interactor.SchedulerSwitcher;
import com.github.il4enkodev.househeating.domain.interactor.UseCaseObservable;
import com.github.il4enkodev.househeating.domain.repository.MeteringRepository;

import javax.inject.Inject;

import io.reactivex.Observable;

public class CurrentMetering extends UseCaseObservable<Metering<Double>, Void> {

    private final MeteringRepository meteringRepository;

    @Inject
    CurrentMetering(SchedulerSwitcher<Metering<Double>> schedulerSwitcher,
                    MeteringRepository meteringRepository) {
        super(schedulerSwitcher);
        this.meteringRepository = meteringRepository;
    }

    @Override
    protected Observable<Metering<Double>> source(Void arguments) {
        return meteringRepository.current();
    }
}

File: domain/src/main/java/com/github/il4enkodev/househeating/domain/interactor/metering/CurrentMetering.kt
package com.github.il4enkodev.househeating.domain.interactor.metering

import com.github.il4enkodev.househeating.domain.di.IoDispatcher
import com.github.il4enkodev.househeating.domain.entity.metering.Metering
import com.github.il4enkodev.househeating.domain.interactor.FlowUseCase
import com.github.il4enkodev.househeating.domain.repository.MeteringRepository
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject

@ExperimentalCoroutinesApi
class CurrentMetering @Inject internal constructor(
    @IoDispatcher private val dispatcher: CoroutineDispatcher,
    private val meteringRepository: MeteringRepository
) : FlowUseCase<Nothing, Metering<Double>>(dispatcher) {

    override suspend fun execute(arguments: Nothing): Flow<Metering<Double>> {
        return meteringRepository.current()
    }
}